File: 프로그래머스/unrated/172928. 공원 산책/공원 산책.py
def solution(park, routes):
    dr = [-1, 0, 1, 0]
    dc = [0, 1, 0, -1]
    dic = {'N':0, 'E':1, 'S':2, 'W':3}
    
    n = len(park)
    m = len(park[0])
    
    # 시작 지점 찾기
    for i in range(len(park)):
        for j in range(len(park[0])):
            if park[i][j] == 'S':
                r, c = i, j

    # 명령어 하나씩 for문 돌기
    for cmd in routes:
        # 방향과 거리 분할 입력
        direction, distance = cmd.split()
        
        # 명령 실행 전 좌표 저장(이동 불가 시 되돌리기 위함)
        nr, nc = r, c
        flag = True
        
        for i in range(int(distance)):
            nr, nc = nr + dr[dic[direction]], nc + dc[dic[direction]]
            if nr < 0 or nr >= n or nc < 0 or nc >= m or park[nr][nc] == 'X':
                flag = False
                break
        if flag:
            r, c = nr, nc

    return [r, c]

File: 프로그래머스/1/172928. 공원 산책/공원 산책.java
import java.util.*;
class Solution {
    int[] dr = {-1, 0, 1, 0};
    int[] dc = {0, 1, 0, -1};
    
    Map<String, Integer> map = new HashMap<>() {{
        put("N", 0);
        put("E", 1);
        put("S", 2);
        put("W", 3);
    }};
    
    public int[] solution(String[] park, String[] routes) {
        
        int[] answer = new int[2];
        int r = 0, c = 0;
        int n = park.length;
        int m = park[0].length();
        
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(park[i].charAt(j) == 'S'){
                    r = i;
                    c = j;
                    for(String route : routes){
                        String[] temp = route.split(" ");
                        
                        int nr = 0, nc = 0;
                        boolean flag = true;
                        
                        for(int k = 1; k <= Integer.parseInt(temp[1]); k++){
                            nr = r + dr[map.get(temp[0])] * k;
                            nc = c + dc[map.get(temp[0])] * k;  
                            
                            if(nr < 0 || nr >= n || nc < 0 || nc >= m || park[nr].charAt(nc) == 'X'){
                                flag = false;
                            }
                        }
                        
                        r = flag ? nr : r;
                        c = flag ? nc : c;
                    }
                }
            }
        }
        return new int[]{r, c};
    }
}